iotilecore/iotile/core/hw/debug/debug_manager.py

```diff
@@ -49,12 +50,145 @@ def dump_ram(self, out_path, pause=False):
+        have a concept of restoring a snapshot of their internal state.
+
+        For those devices this method takes a path to a previously produced
+        snapshot file from a call to save_snapshot() and will load that
+        snapshot into the currently emulated device.
+
+        Args:
+            in_path (path): The output path at which to save
+                the binary core dump.  This core dump consists
+                of the current contents of RAM for the device.
+        """
+
+        with open(in_path, "r") as infile:
+            internal_state = json.load(infile)
+
+        self.restore_snapshot(internal_state)
+
+    def dump_snapshot(self):
+        """Get the current state of the emulated device.
+
+        This debug routine is only supported for emulated devices that
+        have a concept of taking a snapshot of their internal state.
+
+        For those devices this will return a snapshot dictionary with
+        the entire internal state of the device.
+
+        Returns:
+            dict: The internal snapshot of the device's current state.
+        """
+
+        return self._stream.debug_command('dump_state')
+
+    def restore_snapshot(self, snapshot):
+        """Restore a previous state snapshot from an emulated device.
+
+        Args:
+            snapshot (dict): A snapshot of the internal state of this
+                device previously obtained by calling dump_snapshot()
+        """
+
+        self._stream.debug_command('restore_state', {'snapshot': snapshot})
+
+    @docannotate
+    def open_scenario(self, scenario_path):
+        """Load a test scenario from a file into an emulated device.
+
+        This debug routine is only supported for emulated devices and will
+        cause the connected device to load one of its preconfigured test
+        scenarios.  If the scenario has arguments, the args dict
+        may be passed to configure it.
+
+        Args:
+            scenario_path (str): The path to a file containing the test
+                scenario details.
+        """
+
+        with open(scenario_path, "r") as infile:
+            scenario = json.load(infile)
+
+        self.load_scenario(scenario)
+
+    def load_scenario(self, scenario_data):
+        """Load the given test secnario onto an emulated device.
+
+        This debug routine is only supported for emulated devices and will
+        cause the connected device to load one of its preconfigured test
+        scenarios.  If the scenario has arguments, the args dict
+        may be passed to configure it.
+
+        Args:
+            scenario_data (list or dict): Either a dict describing the scenario
+                to be loaded or a list of such dicts that will all be loaded.
+        """
+
+        self._stream.debug_command('load_scenario', {'scenario': scenario_data})
+
+    @docannotate
+    def track_changes(self, enabled=True):
+        """Start of stop tracking all internal state changes made to an emulated device.
```

Comment:
typo -- assume you meant "start or stop"

Reply:
Thanks for flagging that.  Fixed.